### DISCUSSION TITLE
TypeScript: way to reference model type by name ? (`Prisma.*Delegate`)

### DISCUSSION THREAD

**Discussion by haltcase** [prisma/prisma#16058] [Q&A] [answered]
I'm looking for a way to retrieve the type of a given model when I've already got references to both the `PrismaClient` and a `Prisma.ModelName`. My current use case is for middleware that could act on any model, but this could apply anywhere.

```ts
export const testCase = <
	T extends PrismaClient,
	ModelName extends Prisma.ModelName, // e.g., `User`
	FieldName extends keyof typeof Prisma[`${ModelName}ScalarFieldEnum`] // e.g., `email`
>(props: {
	client: T;
	modelName: ModelName;
	sourceField: FieldName;
}): void => {
	// naive camelCase, e.g., User --> user
	const modelReferenceName =
		props.modelName[0].toLowerCase() + props.modelName.substring(1);

	// is there any way to do this without losing type safety?
	const model = (props.client as any)[modelReferenceName];

	// ... aside from hardcoding to a specific model:
	(model as typeof props.client.user).findFirst({
		select: {
			[props.sourceField]: true
		}
	});
};
```

It would be great to have 2 utilities:

1. Compile-time type to map a `Prisma.ModelName` to a model delegate:

  ```ts
  type UserModel = Prisma.Prisma__ModelDelegate<"User">;
  // vs.
  type UserModelHardcoded = typeof client.user;
  
  // both types return `Prisma.UserDelegate<Prisma.RejectOnNotFound | Prisma.RejectPerOperation | undefined>`
  ```

2. Runtime function to get a reference to a model by name instead of having to manually camelCase a given string, e.g.:

  ```ts
  const user = client.$getModel("User"); // returns equivalent of `client.user`
  user.findFirst({ /* ... */ });
  ```
  
These would make it much easier to create type-safe middlewares and other sorts of extensions. I suppose this is similar to https://github.com/prisma/prisma/issues/5273, but I'm not necessarily looking to _extend_ types, just get references to them (though you might get both abilities at once).

**Comment by pantharshit00** [comment, marked as answer]
Hello @haltcase 

Actually we have a similar request like this here: https://github.com/prisma/prisma/issues/6980. So above is not quite possible right now. I will suggest you follow the issue.

**Reply by haltcase** [reply]
Will do, thanks for pointing that one out. 👍

**Comment by r3lays** [comment]
Hey @haltcase did you ever get anywhere with this? I checked out #6980 and didn't really get far with it. 
Seems like the functionality works fine - but obviously the type checking is being a nuisance. 
Seemed like my only solution was to throw a @ts-ignore at it and call it a day but I'd rather not. 

Think you could share any solution or progress you made regarding this?! I'd appreciate it!

**Reply by haltcase** [reply]
@r3lays I didn't. I'm still using my untyped middleware because it worked and I wanted to move on, but I'd still like to do better on the next one. It still seems to me like Prisma could easily just generate more type definitions to represent these things ¯\\\_(ツ)_/¯

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
